storage/zynq7000-flash/flashdrv.c

```diff
@@ -398,126 +357,424 @@ ssize_t flashdrv_read(addr_t offs, void *buff, size_t len, time_t timeout)
-		res = flashdrv_sync();
 		if (res < 0) {
 			qspi_deinit();
 			break;
 		}
 
 		res = qspi_deinit();
-		if (res < 0)
+		if (res < 0) {
 			break;
+		}
 	} while (0);
 
-	free(fdrv_common.buff);
-
 	return res;
 }
 
 
-int flashdrv_init(void)
+int flashdrv_devInit(storage_t *strg)
 {
-	int i, res;
-	size_t sectorSz = 0;
-	flash_info_t *info = &fdrv_common.info;
+	int res;
+	flash_reg_t *reg;
+	size_t secSz;
 
-	fdrv_common.regID = (uint32_t)-1;
-	fdrv_common.sectID = (uint32_t)-1;
-	fdrv_common.buffPos = 0;
+	int id = fdrv_common.initRegs;
+	flash_info_t *info = &fdrv_common.info;
 
-	res = qspi_init();
-	if (res < 0)
+	/* Initialize region data */
+	reg = &fdrv_common.regs[id];
+	res = mutexCreate(&reg->lock);
+	if (res < 0) {
 		return res;
+	}
 
-	res = flashdrv_cfiRead(&info->cfi);
-	if (res < 0)
-		return res;
+	secSz = CFI_SIZE_SECTION(info->cfi.regs[id].size);
+	reg->buffPos = 0;
+	reg->sectID = (uint32_t)-1;
+	reg->buff = malloc(secSz);
+	if (reg->buff == NULL) {
+		resourceDestroy(reg->lock);
+		return -ENOMEM;
+	}
 
-	res = flashcfg_infoResolve(info);
-	if (res < 0)
-		return res;
+	/* Initialize dev structure for new region */
+	strg->dev = malloc(sizeof(storage_dev_t));
+	if (strg->dev == NULL) {
+		resourceDestroy(reg->lock);
+		free(reg->buff);
+		return -ENOMEM;
+	}
 
-	for (i = 0; i < info->cfi.regsCount; ++i) {
-		if (CFI_SIZE_SECTION(info->cfi.regs[i].size) > sectorSz)
-			sectorSz = CFI_SIZE_SECTION(info->cfi.regs[i].size);
+	/* Initialize storage device properties */
+	strg->parent = NULL;
+	strg->start = flashdrv_regStart(id);
+	strg->size = CFI_SIZE_REGION(info->cfi.regs[id].size, info->cfi.regs[id].count);
+
+	/* NOR flash driver supports MTD interface */
+	strg->dev->mtd = malloc(sizeof(storage_mtd_t));
+	if (strg->dev->mtd == NULL) {
+		resourceDestroy(reg->lock);
+		free(reg->buff);
+		free(strg->dev);
+		return -ENOMEM;
 	}
 
-	fdrv_common.buff = (uint8_t *)malloc(sectorSz);
-	if (fdrv_common.buff == NULL)
+/* In case of non-standard erase size, warning information is printed */
+#ifndef NDEBUG
+	if (secSz > MTD_DEFAULT_ERASESZ) {
+		printf("flashdrv: non-standard erase size, file system image properties have to be changed to suit greater size.");
+	}
+#endif
+
+	strg->dev->id = id;
+	strg->dev->mtd->ops = &mtdOps;
+	strg->dev->mtd->type = mtd_norFlash;
+	strg->dev->mtd->name = info->name;
+	strg->dev->mtd->metaSize = 0;
+	strg->dev->mtd->metaAvail = 0;
+	strg->dev->mtd->writesz = 0x1;
+	strg->dev->mtd->writeBuffsz = CFI_SIZE_PAGE(info->cfi.pageSize);
+	strg->dev->mtd->erasesz = secSz < MTD_DEFAULT_ERASESZ ? MTD_DEFAULT_ERASESZ : secSz;
+
+
+	/* NOR flash supports block device interface */
+	strg->dev->blk = malloc(sizeof(storage_blk_t));
+	if (strg->dev->blk == NULL) {
+		resourceDestroy(reg->lock);
+		free(reg->buff);
+		free(strg->dev);
+		free(strg->dev->mtd);
 		return -ENOMEM;
+	}
+	strg->dev->blk->ops = &blkOps;
 
-	return EOK;
+	fdrv_common.initRegs++;
+
+	return (info->cfi.regsCount - fdrv_common.initRegs);
```

Comment:
Isn't some lock needed here?

Reply:
Access to `flashdrv_init` shouldn't be done by many threads, this function has to be invoked only once.
This is the main reason why I decided to not add additional lock for `fdrv_common` fields, in runtime, there are not changed.